graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/migration/CassandraMigrator.java

```diff
@@ -0,0 +1,174 @@
+import io.stargate.db.schema.Table;
+import io.stargate.db.schema.UserDefinedType;
+import io.stargate.graphql.schema.schemafirst.migration.CassandraSchemaHelper.Difference;
+import io.stargate.graphql.schema.schemafirst.processor.EntityMappingModel;
+import io.stargate.graphql.schema.schemafirst.processor.MappingModel;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+public class CassandraMigrator {
+
+  private final DataStore dataStore;
+  private final MappingModel mappingModel;
+  private final MigrationStrategy strategy;
+
+  public CassandraMigrator(
+      DataStore dataStore, MappingModel mappingModel, MigrationStrategy strategy) {
+    this.dataStore = dataStore;
+    this.mappingModel = mappingModel;
+    this.strategy = strategy;
+  }
+
+  /** @throws GraphqlErrorException if the CQL data model can't be migrated */
+  public List<MigrationQuery> compute() {
+
+    Schema schema = dataStore.schema();
+
+    SortedSet<MigrationQuery> queries = new TreeSet<>();
+    List<String> errors = new ArrayList<>();
+
+    for (EntityMappingModel entity : mappingModel.getEntities().values()) {
+      switch (entity.getTarget()) {
+        case TABLE:
+          Table expectedTable = entity.getTableCqlSchema();
+          Table actualTable = schema.keyspace(entity.getKeyspaceName()).table(entity.getCqlName());
+          compute(
+              expectedTable,
+              actualTable,
+              CassandraSchemaHelper::compare,
+              CassandraSchemaHelper::buildCreateQuery,
+              CassandraSchemaHelper::buildDropQuery,
+              CassandraSchemaHelper::buildAddColumnQuery,
+              "table",
+              queries,
+              errors);
+          break;
+        case UDT:
+          UserDefinedType expectedType = entity.getUdtCqlSchema();
+          UserDefinedType actualType =
+              schema.keyspace(entity.getKeyspaceName()).userDefinedType(entity.getCqlName());
+          compute(
+              expectedType,
+              actualType,
+              CassandraSchemaHelper::compare,
+              CassandraSchemaHelper::buildCreateQuery,
+              CassandraSchemaHelper::buildDropQuery,
+              CassandraSchemaHelper::buildAddColumnQuery,
+              "UDT",
+              queries,
+              errors);
+          break;
+        default:
+          throw new AssertionError("Unexpected target " + entity.getTarget());
+      }
+    }
+    if (!errors.isEmpty()) {
+      throw GraphqlErrorException.newErrorException()
+          .message(
+              String.format(
+                  "The schema you provided can't be mapped to the current CQL data model."
+                      + "Consider using a different migration strategy (current: %s). "
+                      + "See details in `extensions.migrationErrors` below.",
```

Comment:
I don't see what does it mean below in this context
`See details in `extensions.migrationErrors` below.`
Will those details be printed out? If yes, where?

Reply:
This should be clearer on an example: create an empty keyspace named `accounts`, then go to the GraphQL playground (`http://$STARGATE_HOST:8080/playground`), enter `http://$STARGATE_HOST:8080/graphqlv2/admin` in the URL bar and execute the following mutation:
```graphql
mutation {
  deploySchema(
    namespace: "accounts"
    migrationStrategy: USE_EXISTING
    schema: """
    type User @key(fields: "id") @cql_input {
      id: ID!
      name: String
      username: String
    }
    type Query {
      getUser(id: ID!): User
    }
    """
  ) {
    version
  }
}
```
Notice the `USE_EXISTING` strategy, which means this will fail because the table does not exist yet.

Here's what the error response will look like:
```graphql
{
  "errors": [
    {
      "message": "Exception while fetching data (/deploySchema) : The schema you provided can't be mapped to the current CQL data model.Consider using a different migration strategy (current: USE_EXISTING). See details in `extensions.migrationErrors` below.",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "deploySchema"
      ],
      "extensions": {
        "migrationErrors": [
          "Missing table User"
        ],
        "classification": "DataFetchingException"
      }
    }
  ],
  "data": {
    "deploySchema": null
  }
}
```